### PR TITLE
appgate-sdp: init at 5.1.2

### PR DESCRIPTION
**Pull request by ymatsiuk**
###### Motivation for this change
Add appgate-sdp vpn client

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+, dnsmasq
+, libappindicator-gtk3
+}:
+with stdenv.lib;
+let
+  deps = [ alsaLib at-spi2-atk at-spi2-core atk cairo cups curl dbus expat gdk-pixbuf glib gtk3 libuuid lttng-ust libsecret nspr nss pango python37 python37Packages.dbus-python libX11 libXScrnSaver libXcomposite libXcursor libXdamage libXext libXfixes libXi libXrandr libXrender libXtst libxcb stdenv.cc.cc libappindicator-gtk3 ];
+  rpath = makeLibraryPath deps + ":${stdenv.cc.cc.lib}/lib64";
+in
+stdenv.mkDerivation rec {
+  name = "appgate-sdp-${version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:61]:
```suggestion
  pname = "appgate-sdp";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please give autoPatchElfHook a try.

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -10174,6 +10174,16 @@
     github = "ymarkus";
     githubId = 62380378;
   };
+  ymatsiuk = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10177]: Please put this into a separate commit with the message ``maintainers: add ymatsiuk``.

File: nixos/modules/programs/appgate-sdp.nix [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    boot.kernelModules = [ "tun" ];
+
+    systemd.services = {
+      appgatedriver = {
+        description = "AppGate driver service";
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          Restart = "always";
+          Type = "forking";
+          ExecStartPre = ''
```
Comment (aanderse) [inline nixos/modules/programs/appgate-sdp.nix:26]: Maybe you want to move this into `preStart`?

File: nixos/modules/programs/appgate-sdp.nix [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            /bin/sh -c "test -e /etc/resolv.appgate && (${pkgs.e2fsprogs}/bin/chattr -i /etc/resolv.conf || :; ${pkgs.coreutils}/bin/mv /etc/resolv.appgate /etc/resolv.conf) ||:"
+          '';
+          ExecStart = "${pkgs.appgate-sdp}/opt/appgate/appgate-driver";
+          PrivateTmp = true;
+          Environment = "PATH=$PATH:${makeBinPath [ pkgs.iproute pkgs.networkmanager pkgs.dnsmasq ]}";
```
Comment (aanderse) [inline nixos/modules/programs/appgate-sdp.nix:31]: Maybe you want to use `paths` instead?

**Review by aanderse — COMMENTED** [review]
Does this package include `systemd` service units? **If** so you might consider using `systemd.packages`.

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (ymatsiuk) [reply maintainers/maintainer-list.nix:10177]: thanks!

File: pkgs/applications/networking/appgate-sdp/default.nix (hunk 1 above)
Comment (ymatsiuk) [reply pkgs/applications/networking/appgate-sdp/default.nix]: I would love to use `autoPatchlfHook` instead, but it doesn't work for this particular package. After it succeeded I see some libraries with missing rpath, which prevents app from starting. I did refine the whole `patchelf` and simplified it.

File: nixos/modules/programs/appgate-sdp.nix (hunk 3 above)
Comment (ymatsiuk) [reply nixos/modules/programs/appgate-sdp.nix:26]: Indeed it does include unit files, but I'm not sure how to wrap those.

**Comment by ymatsiuk** [comment]
@aanderse @SuperSandro2000 guys, please take another look into this PR.
what I've done here is:
* wrap binaries instead of manually adding systemd unit files
* use unit files from `.deb`
* iterate over files in loop and `substituteInPlace`
* set `rpath` for all `*.so` files
* `patchelf` required binaries
* cleaned up dependencies list (left only required packages)

the only concern I have is `substituteInPlace` for some files causes warnings - not sure if that's something anyone cares about?
example:
```
for file in file1 file2 file3; do
  substituteInPlace $file \
    --replace a $out/a \
    --replace b $out/b \
    --replace c $out/c \
    --replace d $out/d
done
```
if `file1` doesn't contain `a` and `file2` `b` I'll get 2 warnings stating the same. Otherwise I would have to expand the loop. What is the preferred way?

File: nixos/modules/programs/appgate-sdp.nix [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      enable = mkEnableOption
+        "AppGate SDP VPN client";
+    };
+  };
+
+  config = mkIf config.programs.appgate-sdp.enable {
+    boot.kernelModules = [ "tun" ];
+    environment.systemPackages = [ pkgs.appgate-sdp ];
+    services.dbus.packages = [ pkgs.appgate-sdp ];
+    systemd.packages = [ pkgs.appgate-sdp ];
```
Comment (aanderse) [inline nixos/modules/programs/appgate-sdp.nix:17]: Awesome! :tada: 

Unfortunately there is [bug](https://github.com/NixOS/nixpkgs/issues/81138) in NixOS... so additional to this you should add something like this:

```
systemd.services.appgate-dumb-resolver.wantedBy =  [ "multi-user.target" ];
systemd.services.appgatedriver.wantedBy =  [ "multi-user.target" ];
systemd.services.appgate-dumb-resolver.wantedBy =  [ "multi-user.target" ];
systemd.services.appgate-resolver.wantedBy =  [ "multi-user.target" ];
```

Comment (ymatsiuk) [reply nixos/modules/programs/appgate-sdp.nix:17]: it does work for me though, but I'm going to give it another test

Comment (aanderse) [reply nixos/modules/programs/appgate-sdp.nix:17]: Oh... Do you mind commenting in the issue thread that you don't have that problem? More data points.

Comment (ymatsiuk) [reply nixos/modules/programs/appgate-sdp.nix]: You are 100% correct. `[Install]` section was present but the unit remained in `Inactive` state. I added suggested fix. Thanks!

Comment (aanderse) [reply nixos/modules/programs/appgate-sdp.nix]: Ah ok. This module looks great!

**Comment by SuperSandro2000** [comment]
> the only concern I have is `substituteInPlace` for some files causes warnings - not sure if that's something anyone cares about?
> example:

I think this is fine for now.

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    stdenv.cc.cc
+    systemd
+    zlib
+  ];
+  rpath = stdenv.lib.makeLibraryPath deps ;
+in
+stdenv.mkDerivation rec {
+  pname = "appgate-sdp";
+  version = "5.1.2";
+  src = fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:96]:
```suggestion
  version = "5.1.2";

  src = fetchurl {
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    url = "https://bin.appgate-sdp.com/5.1/client/appgate-sdp_${version}_amd64.deb";
+    sha256 = "0v4vfibg1giml3vfz2w7qypqzymvfchi5qm6vfagah2vfbkw7xc2";
+  };
+  dontConfigure = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:100]:
```suggestion
  };

  dontConfigure = true;
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  dontBuild = true;
+  enableParallelBuilding = true;
+  buildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:103]:
```suggestion
  enableParallelBuilding = true;

  buildInputs = [
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    python37
+    python37Packages.dbus-python
+  ];
+  nativeBuildInputs = [
+    makeWrapper
+    dpkg
+  ];
+  unpackPhase = ''
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:111]:
```suggestion
  ];

  unpackPhase = ''
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    sha256 = "0v4vfibg1giml3vfz2w7qypqzymvfchi5qm6vfagah2vfbkw7xc2";
+  };
+  dontConfigure = true;
+  dontBuild = true;
+  enableParallelBuilding = true;
+  buildInputs = [
+    python37
+    python37Packages.dbus-python
+  ];
+  nativeBuildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:107]:
```suggestion
  ];

  nativeBuildInputs = [
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    makeWrapper
+    dpkg
+  ];
+  unpackPhase = ''
+    dpkg-deb -x $src $out
+  '';
+  installPhase = ''
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:114]:
```suggestion
  '';

  installPhase = ''
```

File: pkgs/applications/networking/appgate-sdp/default.nix [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        --replace "pkill" "${procps}/bin/pkill"
+    done
+
+    substituteInPlace $out/opt/appgate/linux/nm.py --replace "/usr/sbin/dnsmasq" "${dnsmasq}/bin/dnsmasq"
+    substituteInPlace $out/opt/appgate/linux/set_dns \
+      --replace "service appgate-resolver stop" "${systemd.out}/bin/systemctl stop appgate-resolver" \
+      --replace "/etc/appgate.conf" "$out/etc/appgate.conf"
+
+  '';
+  postFixup = ''
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/appgate-sdp/default.nix:142]:
```suggestion
  '';

  postFixup = ''
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108402` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>appgate-sdp</li>
</ul>
</details>

**Review by aanderse — APPROVED** [review]
The module LGTM 👍
